🐈‍⬛ 2단계 - 로그인 리팩터링

parkjun5 opened the pull request:
안녕하세요!

로그인 리팩터링 을 마치고 리뷰를 요청드립니다.

로그인 리펙터링하면서 RoleBaseAccess 처리하는 인터셉터와
구현에서 요청한 HandlerMethodArgumentResolver 두개로 구현하였습니다.

진행하면서 똑같은 도메인을 이용하지만, 
Admin과 사용자가 요청 방식이 달라 나눠진 코드가 만들어지었는데요.
이런 경우 Application Layer까지 분리를 해줘야하는지, 어디까지 분리하고 어디서부터 공통 부븐으로 처리해주면 좋을지 고민이 많이 되는 미션이였습니다.

리뷰어님은 어떻게 생각하시나요??

이번 리뷰도 잘 부탁드립니다!!

parkjun5 committed:
refactor: test의 경우 port 변경

parkjun5 committed:
refactor: check, count, findId 등 중복 체크 혹은 기존 값 확인 로직 통일

 - FindAnyIdById로 변경 (Id 인덱스 조회로 좀더 빠른 조회)

parkjun5 committed:
feat: 로그인 리팩터링

  - MemberRoleAccessInterceptor를 통한 유저에게 해당 권한이 있는 지 확인
  - LoginMemberArgumentResolver를 통해 헤더에서 값을 불러와준다.

parkjun5 committed:
feat: NeedMemberRole을 통한 권한 확인 추가

 - @NeedMemberRole 를 통한 권한 확인 추가

parkjun5 committed:
feat: 관리자와 일반 사용자의 예약을 분리해주기 위한 ReservationAdminController 분리

parkjun5 committed:
refactor: ReservationDateChecker.validateDate로 통합

parkjun5 committed:
refactor: Token의 이메일 값으로 맴버를 가져오는 로직 추가

parkjun5 committed:
refactor: 테스트 클래스 의존성 변경

parkjun5 committed:
refactor: 2단계 적용을 위한 프런트 변경 및 로직 변경

parkjun5 committed:
refactor: 예약을 관리자 모드로 저장/찾아오기 기능 추가

parkjun5 committed:
polishing

parkjun5 committed:
test: 권한이 필요한 Presentation Layer Test에 쿠키 값 추가

NewWisdom commented on src/main/java/roomescape/support/handler/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,33 @@
+import roomescape.apply.auth.ui.dto.LoginMember;
+import roomescape.support.WebRequestTokenFinder;
+
+public record LoginMemberArgumentResolver(
```
해당 클래스를 record 로 선언한 것은 의도된 것인가요~? 
ThemeController 도 동일하네요!

parkjun5 replied:
인텔리 제이 추천으로 변경한 부분인데, 내부가 불변이라 어울리지 않을까 해서 적용해보았습니다.
https://youtrack.jetbrains.com/issue/IDEA-252036
해당 내용도 제가 최근 버전에서는 고쳐진 내용인 듯 합니다.

그런데 다시 생각해보니까 record에는 필요 없는 함수(hashcode, equals)들도 생기다보니 컨트롤러가 record로 되어 있는 게 썩 합리적이지는 않은 것 같네요 😅

NewWisdom commented on src/main/java/roomescape/support/exception/GlobalExceptionHandler.java:
```diff
@@ -27,7 +27,7 @@ public class GlobalExceptionHandler {
     @ExceptionHandler({TokenNotFoundException.class, IllegalTokenException.class})
     public ResponseEntity<ExceptionResponse> handleTokenException(IllegalArgumentException e) {
-        return ResponseEntity.badRequest().body(new ExceptionResponse(400, e.getMessage()));
+        return ResponseEntity.badRequest().body(new ExceptionResponse(401, e.getMessage()));
```
HTTP 상태코드가 400이 아닌 401 이 되어야 하지 않을까요~? 
추가로 `@ResponseStatus` 을 사용해 볼 수도 있을 것 같네요 :)

parkjun5 replied:
GlobalHandler에 개선점이 많네요.

상태코드를 전반적으로 수정하고, 커스텀 상태코드를 몇가지 추가하였으며, 좀더 간결하게 변경해보겠습니다.

이 예외 처리 부분은 처음부터  잘 잡고 넘어 가야 하는데 깊게 관심 가지지 못해서 언제나 에러가 발생한 뒤에나 고쳐지곤 하네요 ㅎㅎ..

NewWisdom commented on src/main/java/roomescape/support/handler/MemberRoleAccessInterceptor.java:
```diff
@@ -0,0 +1,79 @@
+import static java.util.stream.Collectors.toSet;
+
+@Component
+public class MemberRoleAccessInterceptor implements HandlerInterceptor {
```
> 3단계 요구사항: Member의 Role이 ADMIN 인 사람만 /admin 으로 시작하는 페이지에 접근할 수 있습니다.

현재 게스트 권한으로도 어드민 페이지 접근이 가능한 것 같은데 확인 부탁드릴게요~!

NewWisdom replied:
권한 확인에 Interceptor을 사용하는 것이 미션 요구사항이라 자연스럽게 HandlerInterceptor 를 사용했을 것 같은데요, 
스프링이 제공하는 filter 를 활용할 수도 있었을 것 같은데 Interceptor 와 filter 의 차이는 무엇인가요~? 
왜 interceptor 가 적합했는지 정리하면 좋을 것 같아요 ㅎㅎ

parkjun5 replied:
RestContoller에만 권한 적용하다보니 페이지 조회를 담당하는 컨트롤러에는 빠져있었군요!

더 추가로 적용해주겠습니다.
또 하다보니...401에러가 나는게 아니라 login 페이지로 가야겠네요.
또 login하면 리다이렉트가 되면 더 좋겠고요.. 할게 느네요 ㅎㅎㅎ

parkjun5 replied:
인터셉터와 필터의 차이는

인터셉터는 요청이 올 때 컨트롤러 전 후에서 검사나 로깅같이 로직을 넣어 줄 때에 주로 사용하고 있습니다.
경로 패턴에 설정할 수 있어서 좀 더 적용이 편해서 많이 사용하는 듯 합니다.
필터랑 다른 점은 리턴 타입이 true or false여서 잘못되었거나 허용되지 않는 요청을 막아주는 역할을 합니다.

필터는 서버와 클라이언트 간의 요청과 응답을 수정할 수 있다는 점이 다른거 같아요.
스프링 디스페처서블릿을 통해 작동되던 인터셉터와 달리 스프링과는 상관없이 모든 자원에 적용이 가능해요.
서블릿 요청뿐만아니라 다른 자원에도 적용이 가능해서 요청 자체의 값을 바꿔줄수도 있고 응답값의 인코딩을 바꿔주는데 주로 사용해요.

간단히 요약하면 요청을 막을 떄에는 인터셉터, 
요청과 응답을 변환하면서 도와주는 친구는 필터가 아닐까 합니다.

NewWisdom replied:
잘 정리해주셨네요 👍

NewWisdom commented on src/main/java/roomescape/support/handler/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,33 @@
+                                  NativeWebRequest webRequest,
+                                  WebDataBinderFactory binderFactory
+    ) {
+        String token = WebRequestTokenFinder.getTokenByRequestCookies(webRequest);
```
```
HttpServletRequest request = (HttpServletRequest) webRequest.getNativeRequest();
```
NativeWebRequest, HttpServletRequest 각각을 위한 token 값을 얻는 클래스들이 따로 존재하는 것 같은데 여기서 HttpServletRequest 로 바꿔주면 중복 로직을 공통화 할 수 있겠네요!

parkjun5 replied:
webRequest에서 HttpServletRequest를 가져오면 좋은 것 같아요!

그런데 형병환도 좋은 방법이지만 

`final HttpServletRequest httpServletRequest = webRequest.getNativeRequest(HttpServletRequest.class);`

저는 이런 방식이 조금 더 안전할 것 같아요.

NewWisdom commented on src/main/java/roomescape/apply/auth/application/JwtTokenManager.java:
```diff
@@ -67,4 +72,10 @@ private Claims parseJwt(String token) {
+    public LoginMember getMemberEmailAndNameBy(String token) {
+        String email = parseJwt(token).getSubject();
+        return memberFinder.getLoginMemberByEmail(email);
+    }
```
LoginMember 객체 자체를 반환하는데 네이밍을 보고 유추하기 어려울 것 같네요!

NewWisdom commented on src/main/java/roomescape/support/handler/MemberRoleAccessInterceptor.java:
```diff
@@ -0,0 +1,79 @@
+        Set<MemberRoleName> requiredRoleNames = getRequiredMemberRolesInMethod(handler);
+        boolean containRole = requiredRoleNames.stream().anyMatch(memberRoleNames::contains);
+        if (!containRole) {
+            response.setStatus(HttpServletResponse.SC_FORBIDDEN);
```
403 숫자가 아닌 HttpServletResponse 의 SC_FORBIDDEN 값을 활용 👍

NewWisdom commented on src/main/java/roomescape/apply/member/domain/repository/MemberRepository.java:
```diff
@@ -10,7 +10,11 @@ public interface MemberRepository {
     List<Member> findAll();
 
-    Integer countByEmail(String email);
+    Optional<Long> findAnyIdByEmail(String email);
 
     Member save(Member member);
+
+    Optional<Member> findOneByEmail(String email);
+
+    Optional<Member> findOneById(long memberId);
```
메서드 명에 Any, One을 붙이는 기준은 무엇인가요~? repository 단 메서드 네이밍의 컨벤션을 정리해봐도 좋을 것 같네요 ㅎㅎ

parkjun5 replied:
메서드 명을 만들때,
어떤 것이든 찾으면 되는 것은 Any
정확한 값을 찾을 때에는 One으로 사용합니다.

그런데 유니크한 Email로 찾으니...Any가 의미가 없네요.😂
어차피 한가지 값만 나와야 할테니까요.
아무 이름의 아이디를 찾을 떄에는 아래처럼 하지 않을까 합니다.

`findAnyIdByName(String name);`

NewWisdom commented on src/main/java/roomescape/apply/reservationtime/ui/TimeController.java:
해당 컨트롤러 네이밍도 `ReservationTimeController` 은 어떨까요~?

parkjun5 replied:
이 부분은 고민이 되는 게 도메인 명은 말씀해주신대료 ReservationTime이 맞지만
RequestMapping은 `/times` 이라 고민이 되네요.

만약 컨트롤러의 네이밍을 변경한다면 매핑도 reservationTimes로 변경해 주는 게 맞지 않나 싶습니다.

컨트롤러의 이름은 매핑과 맞춰 주는 게 많이 익숙했던 것 같아서요.

NewWisdom replied:
의견 감사합니다 👍 
현재 프로젝트에서는 컨트롤러만 다른 네이밍을 쓰고 있어서 의도된건지 궁금하기도 하고, 저는 주로 도메인관점에 맞추어서 네이밍을 주로 썼던 것 같아서 여쭈어 보았어요~! 
(조금 다른 이야기이지만 개인적으로는 맵핑도 "/reservation-times"처럼이 조금 더 명확하지 않을까 생각이 들기도 했네요 ㅎㅎ)

NewWisdom commented on src/main/java/roomescape/support/handler/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    @Override
+    public Object resolveArgument(MethodParameter parameter,
```
해당 메서드의 반환 타입을 LoginMember 로 정의하는 것은 어떤가요~?

NewWisdom commented on src/main/java/roomescape/apply/reservation/ui/ReservationController.java:
```diff
@@ -32,12 +35,16 @@ public ResponseEntity<List<ReservationResponse>> getReservations() {
 
     @PostMapping
-    public ResponseEntity<ReservationResponse> addReservation(@RequestBody ReservationRequest request) {
+    @NeedMemberRole({MemberRoleName.ADMIN, MemberRoleName.GUEST})
+    public ResponseEntity<ReservationResponse> addReservation(@RequestBody ReservationRequest request,
+                                                              LoginMember loginMember
+    ) {
         return ResponseEntity.status(HttpStatus.CREATED)
-                             .body(reservationRecorder.recordReservationBy(request));
+                .body(reservationRecorder.recordReservationBy(request, loginMember));
     }
```
> (미션 요구사항) 예약 생성은 로그인 사용자와 관리자 모두 사용할 수 있는 기능입니다.
관리자는 memberId 인자로 전달한 정보로 예약을 생성하게 합니다.
로그인 사용자는 자신의 로그인 정보로 예약을 생성하게 합니다.

> 진행하면서 똑같은 도메인을 이용하지만,
Admin과 사용자가 요청 방식이 달라 나눠진 코드가 만들어지었는데요.
이런 경우 Application Layer까지 분리를 해줘야하는지, 어디까지 분리하고 어디서부터 공통 부븐으로 처리해주면 좋을지 고민이 많이 되는 미션이였습니다.

미션 요구사항과 함께 확인해 보니 말씀 주신 부분이 해당 부분 맞을까요~?
어드민과 사용자의 API 스펙이 다르다면 사실 API 자체를 분리하는 것이 맞을 것이고, 
미션에서 요구한대로 관리자, 로그인 사용자 각각의 정보로 예약 생성이라면,  ArgumentResolver 사용을 의도한 것 같은데 적절히 사용해주신 것 같네요 👍 
혹시 다른 부분을 이야기하신거라면 편히 말씀 주세요 :)

parkjun5 replied:
말씀해주신 부분이 맞습니다.
Reservation 이라는 하나의 도메인에 

어드민에서 요청하는 부분과 게스트에서 요청하는 부분으로 나뉘게 되었어요.

그러면서 Persetation Layer는 자동으로 나뉘게 되었는데, 서비스 레이어도 분리를 해줘야 할까 라는 고민이 생긴 것 기였습니다.

제 생각으로는 아직 서비스 레이어에 중복이 많고 서로 처리가 많이 다르지 않아서, 서비스 레이어에서는 분리 해줄 필요가 없다고 생각했습니다.

나중에 어드민과 게스트의 비지니스 요구사항이 많이 차이가 나고 복잡해진다면 분리를 해줘야 하지 않을까 합니다.

parkjun5 replied:
나중에 변경을 대비하려면 이 서비스 레이어에서는 인터페이스를 추출하거나 중복 로직을 모듈화 시키면 어떨까요??

NewWisdom replied:
👍 
말씀해주신대로 현재는 권한에 따라 같은 API 에 대해 다른 처리를 미션에서 요구하고 있고, 로직 차이가 없어 공통으로 처리해도 문제가 없지만 권한에 따라 다른 동작이 필요하다면 (미션이 아니라면) API 자체를 다르게 분리해야하지 않을까 싶어요 ㅎㅎ  

> 나중에 변경을 대비하려면 이 서비스 레이어에서는 인터페이스를 추출하거나 중복 로직을 모듈화 시키면 어떨까요??

사실 현재 단계에서는 오히려 복잡도를 높일 수도 있으니 현재 구조로도 충분한 것 같네요 !
(소프트웨어 개발 3대 원칙에 YAGNI - You Ain't Gonna Need it 라는 말도 있죠~? 😄)

NewWisdom commented on src/main/java/roomescape/support/handler/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,33 @@
+        String token = WebRequestTokenFinder.getTokenByRequestCookies(webRequest);
+        jwtTokenManager.validateToken(token);
+        return jwtTokenManager.getMemberEmailAndNameBy(token);
+    }
```
jwtTokenManager는 어떤 역할을 하는 객체인가요~?
여기서 네이밍을 보면 jwt 토큰 관리하는 역할을 할 것 같은데 loginMember 까지 만들어주고 있네요!
현재 구조로 보면 여기서 memberFinder에게 바로 loginMember 를 찾게 해줄 수도 있을 것 같은데 어떻게 생각하시나요~?

parkjun5 replied:
TokenManager가 토큰 관리 이상의 역할이 되어 버렸네요.

말씀해주신 방법으로 사용하는게 깔끔한 것 같아요!

NewWisdom commented on src/main/java/roomescape/support/handler/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,33 @@
+        jwtTokenManager.validateToken(token);
+        return jwtTokenManager.getMemberEmailAndNameBy(token);
+    }
+}
```
파일 끝 개행 한번 확인 부탁드려요 ㅎㅎ

NewWisdom commented on src/test/java/roomescape/apply/auth/application/JwtTokenManagerTest.java:
```diff
@@ -1,27 +1,58 @@
-    JwtTokenManagerTest() {
+    @BeforeEach
+    void setUp() {
+        transactionStatus = transactionManager.getTransaction(new DefaultTransactionDefinition());
```
BaseTestService 에서 `@JdbcTest` 를 사용하신 것 같은데 해당 어노테이션이 해주는 기능을 조금 더 확인해보면 좋을 것 같네요!
번외로 `@JdbcTest`와 `@DataJdbcTest` 는 어떤 차이가 있을까요~?

parkjun5 replied:
@JdbcTest에 대해 말씀해주신 이유는 제가 강제로 transaction을 rollback 시키는 부분과 @AutoConfigureTestDatabase 부분 때문 일 것 같아요!

@JdbcTest에서 기본적으로 각 테스트가 트랜잭션 내에서 실행되고 테스트가 끝나면 자동으로 롤백되지만,데이터베이스의 설정에 따라 오토 인크리먼트 필드나 시퀀스가 롤백되지 않은 부분이 있었어요. 특히 MySQL과 같은 일부 데이터베이스에서는 트랜잭션 롤백 시에도 자동 증가 값이 초기화되지 않더라구요. 그래서 테스트를 작성할 때에 일부러 넣어주려고 하고 있어요!

또한 @AutoConfigureTestDatabase는 엔티티에 데이터베이스에 종속적인 코드가 있을 떄에 유용합니다.
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE) 으로 설정하면 제대로 작동하더구요!
그래서 BaseJpaTest를 작성할 때에 기본적으로 추가해주고는 있습니다.

물론 지금 H2 InMemory 환경에서는 다 필요없는 설정이지만요.🥺
그래도 어떤 데이터베이스에서도 동작하는 테스트 코드를 위해서 추가해두는 것이 좋다고 생각했어요.😊

parkjun5 replied:
@JdbcTest와 @DataJdbcTest는 지원하는 데이터베이스 매퍼에 따라 달라요!

JdbcTest에서는 간단한 Jdbc 관련 설정과 JdbcTemplate만 지원합니다. JdbcTemplate으로 쿼리를 직접 작성해서 실행해서 간단한 쿼리는 빠르게 실행할 수 있어서 좋아요.

DataJdbcTest는 Jdbc 설정 이상의 CRUD repository 설정도 지원합니다. 이 CRUD repository에는 CRUD 기능을 자동으로 생성해주어 편리합니다. 또한 인덱스를 관리해주고 쿼리도 자동으로 생성해줘서 편리하죠.

그런데 제 구현 코드에서는 DataJdbc를 사용하고 있지 않아서 @JDBCTest를 설정해주었습니다!

NewWisdom replied:
크 잘 정리해주셨네요 👍

NewWisdom reviewed (changes requested):
세준님 안녕하세요 ~!
2단계도 빠르게 잘 진행해주셨네요 👍 
몇가지 코멘트 남겨놓았는데 확인해보시면 좋을 것 같습니다!
이번 단계도 화이팅입니다 ㅎㅎㅎ

Commit:
refactor: remove duplicate TokenFinder WebRequestTokenFinder

- NativeWebRequest에서 쿠키를 가져오지 말고 HttpServletRequest로 변환하여 기존 TokenFinder를 사용하도록 변경
- 사용하지 않게된 WebRequestTokenFinder 제거

parkjun5 committed:
refactor: rollback change class type controller and resolver

 - 컨트롤러와 리솔버에 record를 적용해본 부분 롤백
 - 인텔리제이 추천 이슈 https://youtrack.jetbrains.com/issue/IDEA-252036

parkjun5 committed:
refactor: redefine HttpStatus in GlobalExceptionHandler

- Fix mismatched status codes
- Added detailed custom status codes
- Simplified return types using @ResponseStatus

parkjun5 committed:
refactor: Apply authorization to IndexController and add login redirect

 - RestController 외에 페이지 조회를 담당하는 컨트롤러에도 권한 적용
 - 권한이 없는 경우 401 에러 대신 로그인 페이지로 리다이렉트
 - 로그인 후 원래 요청했던 페이지로 리다이렉트되는 기능 추가

parkjun5 committed:
refactor: FindAnyId, FindOneId 정리 FindId로 통일

parkjun5 committed:
refactor: JwtTokenManager dont care LoginMember

 - 토큰 매니저의 과도한 책임 분리

parkjun5 committed:
test: Fix Test By Add Check MemberRole To Pages

 - 페이지 접속 권한을 확인하면서 기존 테스트 수정
 - 테스트용 마스터 계정 추가

NewWisdom reviewed (changes requested):
안녕하세요 세준님~ 
피드백 빠르게 잘 반영해주셨네요 👍 
남겨주신 코멘트들도 잘 보았습니다 ! ㅎㅎ 
다만 현재 페이지에서 회원가입이 제대로 되지 않는 것 같은데 
전체적으로 확인 후 다시 리뷰 요청 부탁드릴게요 🙇

Commit:
fix: Fix Signup reloading page Error

 - 가입하면서 재로딩되면서 화면이 안움직이는 오류 수정
 - 회원 가입시에는 권한확인 제외
 - 로그아웃 추가

parkjun5 committed:
refactor: change error Message

 - 에러 메시지에 좀더 상세한 내용을 포함하도록 변경

parkjun5 commented:
권한 적용 도중에 사용자 추가에도 어드민 권한을 요청하게 되면서 문제가 되었네요.

바로 적용하였습니다!

또한 에러메시지도 더 상세하게 보여지도록 적용했어요!

NewWisdom reviewed (approved):
세준님 안녕하세요 ~! 
수정해주신 부분 확인하였습니다!
빠르게 잘 반영해주셨네요 👍 
괜찮으시면 추가 미션도 진행해보아도 좋겠네요 :) 끝까지 화이팅입니다!
